### DISCUSSION TITLE
Unable to exchange token with APPLE's Id token via ADMIN api

### DISCUSSION THREAD

**Discussion by asim46** [keycloak/keycloak#30021] [Q&A]
```
String requestBody = "grant_type=urn:ietf:params:oauth:grant-type:token-exchange"  +
        "&client_id=" + clientId +
        "&subject_issuer=" +  provider+
        "&subject_token=" +  accessToken+
        "&subject_token_type=urn:ietf:params:oauth:token-type:"+tokenType; 
```
        
tokenType:id_token
provider: apple

        Response:
        "body":"400 Bad Request: \"{\"error\":\"invalid_token\",\"error_description\":\"token type not supported\"}\""}}
         [error] - ERROR 400
        
It was working until 22.0.4... Its not working in 24.0.4

Its working with GOOGLE:-

tokenType:access_token
provider:google

**Comment by timbrown5** [comment]
I see similar behaviour with GitHub tokens. Token exchange worked with GitHub ID tokens up until 24.0.2, but 24.0.3 fails with: `token type not supported`.

I believe it is due to [this feature](https://github.com/keycloak/keycloak/commit/0d60e5802915118f9c8a1a42382efcdb4af7ec3d#diff-691739297da98c8304c670423103e7fa0f1bea77862b98f551814aa652522501R51) which tightens the requirements for subject token `typ` headers (as per https://datatracker.ietf.org/doc/html/rfc9068#section-2.1) - the GitHub ID Token `typ` header is just, `JWT`.

As we this is a third-party header, I don't think there is an easy workaround? It would be nice to be able to disable this check and/or set additional allowed token `typ` header values.

I raised a question on GitHub [here](https://github.com/orgs/community/discussions/142868), but not expecting much of a response.

**Reply by timbrown5** [reply]
Just found https://github.com/keycloak/keycloak/issues/33332 which could be related.

**Reply by infl00p** [reply]
Hi @timbrown5, 
the KC implementation is deviating from the RFC, because the typ requirement is for the JWT header and not for the data section. In my opinion the check does not need to validate the token type, because in the OIDC identity provider the validateJwt() already checks if the token type is null in a previous step and it sets it to a default value (access token). There is no reason for the token type to be re-evaluated.
 
Basically,  KC  has a very strange regression, because half of the included social providers break with it. Along with many proprietary and opensource OIDC idp implementations.
